### PR TITLE
[Speculation] Add Operations to enable Speculation

### PR DESCRIPTION
**Pull request by asleix**
Speculative execution involves speculating on the output of a long latency operation or chain of operations, allowing operations that depend on it to begin execution early, without impacting correctness.

This pull-request adds four Operations:
- SpeculatorOp. Drives Speculation process and connects to the other units.
- SpecSaveOp. Marks beginning of speculation path.
- SpecCommitOp. Marks end of speculation path.
- SpecSaveCommitOp. Joint Save and Commit that occurs when backedges (self-loops) are present.

The definitions do not include a custom assembly format. If needed, I can add it and would require some indications on how to do so.

This PR is needed for a parent PR in dynamatic.

### CODE REVIEW

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (lucas-rami) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: nit pick: add empty line at the end of the file

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (lucas-rami) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: I'm assuming these are, respectively, `saveCtrl`, `commitCtrl`, and `scBranchControl`. What are `SCControl1` and `SCControl2` then? Why are they 3 bits? Describing them briefly in this `description` field would be useful. If feasible, also consider using more explicit names, 1 and 2 are not very clear as suffixes.

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (lucas-rami) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: nit pick: non-intrinsic attributes (`name`) should not appear here, delete the attribute dictionary from the example. Same comment applies to other operations.

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (lucas-rami) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: nit: use more descriptive SSA value names than `%22` and `%23` in examples. Same applies to other operation examples.

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (lucas-rami) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: All operations are missing the `assemblyFormat` field. A default implementation may be generated automatically but it's good practice to have the format explicitly specified so that it's easier for the reader to understand the operation's format. I recommend taking inspiration from other Handshake operations to understand MLIR's expected format for this field (otherwise, refer to the [official documentation](https://mlir.llvm.org/docs/DefiningDialects/Operations/#declarative-assembly-format)).

**Review by lucas-rami — COMMENTED** [review]
There are a couple cosmetic things to address but after that it will be good to go!

**Comment by asleix** [comment]
I have addressed @RamirezLucas comments to improve the format and clarity.

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (murphe67) [inline include/circt/Dialect/Handshake/HandshakeOps.td]: I really don't like 'optional'- What you intended here was 'conditional' which is not quite the same thing.

I'd be curious if there was a good way to phrase "save-like" and "commit-like" control signals? Ctrl1 is used to control the saving/reading behavior of the save-commit and Ctrl2 is used to control some of the killing, the commit bit of the save-commit. But it might be too much information to try encode in a variable name.

Comment (lucas-rami) [reply include/circt/Dialect/Handshake/HandshakeOps.td]: Perhaps a custom MLIR type instead of a simple `i3` would be clearer (never done anything like this myself but certainly possible)? Or custom getters on the operation itself to get a specific bit of the i3 operands by "purpose"? Fine by me if we declare this to be "future improvement material", just wanted to mention our possibilities.

**Review by lucas-rami — COMMENTED** [review]
LGTM once @murphe67's optional/conditional comment is addressed.

File: include/circt/Dialect/Handshake/HandshakeOps.td
Comment (murphe67) [reply include/circt/Dialect/Handshake/HandshakeOps.td]: We're constrained to [this kind of a structure](https://sunset-hardware-3ee.notion.site/8ab7584e10bd4e48b0e4f61c2c81891e?pvs=4) until we change the VHDL code, and I think 3 bits with one 'standard' control signal and one conditional control signal does convey that structure the best.  

The VHDL code needs quite a big overhaul, but getting what we had before running with MLIR is the priority before looking at it, so it occasionally forces us into making slightly weird short term decisions.

Comment (asleix) [reply include/circt/Dialect/Handshake/HandshakeOps.td]: I have changed the names of the save-commit control signals as @murphe67 suggested.

**Review by lucas-rami — APPROVED** [review]

**Pull request merged**
